src/webgpu/web_platform/external_texture/video.spec.ts

```diff
@@ -234,64 +263,75 @@ updated. Using expired GPUExternalTexture results in an error.
-        const commandBuffer = useExternalTexture();
-        t.expectGPUError('validation', () => t.device.queue.submit([commandBuffer]), false);
-        t.expect(!externalTexture.expired);
+    await startPlayingAndWaitForVideo(videoElement, async () => {
+      const source =
+        sourceType === 'VideoFrame'
+          ? await getVideoFrameFromVideoElement(videoElement)
+          : videoElement;
+      externalTexture = t.device.importExternalTexture({
+        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+        source: source as any,
       });
-
-      // 2. importExternalTexture which should stay active if video frame is not updated.
-      externalTexture = t.device.importExternalTexture({ source: video });
       // Set `bindGroup` here, which will then be used in microtask1 and microtask3.
       bindGroup = t.device.createBindGroup({
         layout: bindGroupLayout,
         entries: [{ binding: 0, resource: externalTexture }],
       });
 
-      // 3. Enqueue a microtask which uses the GPUExternalTexture. The GPUExternalTexture
-      // should still keep alive.
-      const microtask3 = Promise.resolve().then(() => {
-        const commandBuffer = useExternalTexture();
-        t.expectGPUError('validation', () => t.device.queue.submit([commandBuffer]), false);
-        t.expect(!externalTexture.expired);
-      });
+      const commandBuffer = useExternalTexture();
+      t.expectGPUError('validation', () => t.device.queue.submit([commandBuffer]), false);
+      t.expect(!externalTexture.expired);
 
-      // Now make sure the test doesn't end before all of those microtasks complete.
-      await microtask1;
-      await microtask3;
+      if (sourceType === 'VideoFrame') {
+        (source as VideoFrame).close();
+        const commandBuffer = useExternalTexture();
+        t.expectGPUError('validation', () => t.device.queue.submit([commandBuffer]), true);
+        t.expect(externalTexture.expired);
+      }
     });
 
-    // Update new video frame.
-    await startPlayingAndWaitForVideo(video, async () => {
-      // 4. VideoFrame is updated. GPUExternalTexture should be expired. Using the
-      // GPUExternalTexture should result in an error.
-      const commandBuffer = useExternalTexture();
-      t.expectGPUError('validation', () => t.device.queue.submit([commandBuffer]), true);
-      t.expect(externalTexture.expired);
-    });
+    if (sourceType === 'VideoElement') {
+      // Update new video frame.
+      await startPlayingAndWaitForVideo(videoElement, async () => {
+        // VideoFrame is updated. GPUExternalTexture imported from HTMLVideoElement should be expired.
+        // Using the GPUExternalTexture should result in an error.
+        const commandBuffer = useExternalTexture();
+        t.expectGPUError('validation', () => t.device.queue.submit([commandBuffer]), true);
+        t.expect(externalTexture.expired);
+      });
+    }
```

Comment:
Aside: I'm not sure `startPlayingAndWaitForVideo` will actually guarantee we've reached the next frame. AFAIK it's only designed to make sure the video is playing *at all*, not that it's advanced from where it was. It will probably work when `requestVideoFrameCallback` is available, but it isn't always.

How about using `requestVideoFrameCallback` directly instead of `startPlayingAndWaitForVideo`?

```suggestion
    if ('requestVideoFrameCallback' in videoElement) {
      await new Promise(resolve => {
        // Update new video frame.
        videoElement.requestVideoFrameCallback(() => {
          // VideoFrame is updated. GPUExternalTexture imported from HTMLVideoElement should be expired.
          // Using the GPUExternalTexture should result in an error.
          const commandBuffer = useExternalTexture();
          t.expectGPUError('validation', () => t.device.queue.submit([commandBuffer]), true);
          t.expect(externalTexture.expired);

          resolve(undefined);
        });
      });
    }
```

Reply:
That would be ideal. I also find that only rVFC could ensure the update.